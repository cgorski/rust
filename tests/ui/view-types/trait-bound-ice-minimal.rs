#![feature(view_types)]
#![allow(incomplete_features)]

// Minimal reproduction: Generic function with trait bound calling view-typed trait method
// RESTRICTION: View types are not supported on trait methods - should error clearly, not ICE

trait Update {
    fn update(&{mut value} mut self); //~ ERROR patterns aren't allowed in functions without bodies
    //~| WARN this was previously accepted by the compiler but is being phased out
}

struct Data {
    value: i32,
    other: String,
}

impl Update for Data {
    fn update(&{mut value} mut self) { //~ ERROR view types are not supported on trait method implementations
        self.value += 1;
    }
}

// ICE occurs here: generic function with trait bound
fn use_trait<T: Update>(item: &mut T) {
    item.update();  // Calling trait method with view type through generic
}

fn main() {
    let mut data = Data {
        value: 0,
        other: String::new(),
    };

    use_trait(&mut data);

    println!("Value: {}", data.value);
}
