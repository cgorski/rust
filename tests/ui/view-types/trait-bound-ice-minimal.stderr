error: patterns aren't allowed in functions without bodies
  --> $DIR/trait-bound-ice-minimal.rs:8:15
   |
LL |     fn update(&{mut value} mut self);
   |               ^^^^^^^^^^^^^^^^^^^^^ help: remove `mut` from the parameter: `self`
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #35203 <https://github.com/rust-lang/rust/issues/35203>
   = note: `#[deny(patterns_in_fns_without_body)]` (part of `#[deny(future_incompatible)]`) on by default

error: view types are not supported on trait method implementations
  --> $DIR/trait-bound-ice-minimal.rs:18:15
   |
LL |     fn update(&{mut value} mut self) {
   |               ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: view types are restricted to inherent methods
   = note: trait method support requires trait system integration
   = help: consider using regular references without view type annotations

error: aborting due to 2 previous errors

